junit-platform-engine/src/main/java/org/junit/platform/engine/support/descriptor/UriSource.java

```diff
@@ -33,4 +39,42 @@
 	 */
 	URI getUri();
 
+	/**
+	 * Create a new {@code UriSource} using the supplied {@code uri}.
+	 * <p>
+	 * This implementation tries resolve the {@code uri} to local file
+	 * system path-based source first. If that fails for any reason, an
+	 * instance of a simple default uri source class storing the supplied
+	 * {@code uri} <em>as-is</em> is returned.
+	 *
+	 * @param uri the uri instance; must not be {@code null}
+	 * @return a uri source instance
+	 * @since 1.3
+	 * @see org.junit.platform.engine.support.descriptor.FileSource
+	 * @see org.junit.platform.engine.support.descriptor.DirectorySource
+	 */
+	static UriSource from(final URI uri) {
+		Preconditions.notNull(uri, "uri must not be null");
+		try {
+			URI pathBasedUri = uri;
+			String query = pathBasedUri.getQuery();
+			if (StringUtils.isNotBlank(query)) {
+				String s = pathBasedUri.toString();
+				pathBasedUri = URI.create(s.substring(0, s.indexOf('?')));
+			}
+			Path path = Paths.get(pathBasedUri);
+			if (Files.isRegularFile(path)) {
+				return FileSource.from(path.toFile(), FilePosition.fromQuery(query).orElse(null));
+			}
+			if (Files.isDirectory(path)) {
+				return DirectorySource.from(path.toFile());
+			}
+		}
+		catch (IllegalArgumentException e) {
+			LoggerFactory.getLogger(UriSource.class).debug(e, () -> "uri not path-based: " + uri);
```

Comment:
Please store the logger in a static field.

Reply:
uri not path based --> The supplied URI [%s] is not path-based. Falling back to default UriSource implementation.

Reply:
Oops.  Didn't realize I was commenting on a static factory method in an interface.